raiden/src/store/epics/matrix.ts

```diff
@@ -263,3 +305,416 @@ export const matrixPresenceUpdateEpic = (
+    // but working inside concatMap, which is called only at outer next and subscribe delayed
+    multicast(new ReplaySubject<[Presences, RaidenState]>(1), presencesStateReplay$ =>
+      // actual output observable, handles MessageSendAction serially and create room if needed
+      action$.pipe(
+        ofType<RaidenActions, MessageSendAction>(RaidenActionType.MESSAGE_SEND),
+        // this mergeMap is like withLatestFrom, but waits until matrix$ emits its only value
+        mergeMap(action => matrix$.pipe(map(matrix => ({ action, matrix })))),
+        // concatMap is used to prevent bursts of messages for a given address (eg. on startup)
+        // of creating multiple rooms for same address
+        concatMap(({ action, matrix }) =>
+          // presencesStateReplay$+take(1) acts like withLatestFrom with cached result
+          presencesStateReplay$.pipe(
+            // wait for user to be monitored
+            filter(([presences]) => action.address in presences),
+            take(1),
+            // if there's already a room state for address and it's present in matrix, skip
+            filter(
+              ([, state]) =>
+                !get(state, ['transport', 'matrix', 'address2rooms', action.address, 0]),
+            ),
+            // else, create a room, invite known user and dispatch the respective MatrixRoomAction
+            // to store it in state
+            mergeMap(([presences]) =>
+              matrix.createRoom({
+                visibility: 'private',
+                invite: [presences[action.address].userId],
+              }),
+            ),
+            map(({ room_id: roomId }) => matrixRoom(action.address, roomId)),
+          ),
+        ),
+      ),
+    ),
+  );
+
+/**
+ * Invites users coming online to rooms we may have with them
+ */
+export const matrixInviteEpic = (
+  action$: Observable<RaidenActions>,
+  state$: Observable<RaidenState>,
+  { matrix$ }: RaidenEpicDeps,
+): Observable<RaidenActions> =>
+  action$.pipe(
+    ofType<RaidenActions, MatrixPresenceUpdateAction>(RaidenActionType.MATRIX_PRESENCE_UPDATE),
+    filter(action => action.available),
+    // this mergeMap is like withLatestFrom, but waits until matrix$ emits its only value
+    mergeMap(action => matrix$.pipe(map(matrix => ({ action, matrix })))),
+    withLatestFrom(state$),
+    mergeMap(([{ action, matrix }, state]) => {
+      const roomId: string | undefined = get(state, [
+        'transport',
+        'matrix',
+        'address2rooms',
+        action.address,
+        0,
+      ]);
+      if (roomId) {
+        const room = matrix.getRoom(roomId);
+        if (room) {
+          const member = room.getMember(action.userId);
+          if (!member) return from(matrix.invite(roomId, action.userId));
+        }
+      }
+      return EMPTY;
+    }),
+    ignoreElements(),
+  );
+
+/**
+ * Handle invites sent to us and accepts them iff sent by a monitored user
```

Comment:
nitpick: iff/if

Reply:
@kelsos `iff` is an abbreviation for `if and only if`, more commonly used on math, but also valid on algorithms context, like this one